### PR TITLE
feat: add retry logic for k8s client #7692

### LINKED ISSUE #7692: Application controller should be more resilient to network latency or K8s API server hiccups
# Summary

On network links with a very high latency, or which are very slow, sometimes a remote Kubernetes API drops the connection before a certain request (e.g. fetching remote resources, list of available APIs, etc) could be completed. This currently leads to comparison errors in the application or in aborting other operations (e.g. retrieving manifests using `argocd app manifests`). Effectually, this can prevent a sync from happening when such a hiccup occurs within a given operation, because the errors are considered fatal. 

This can be observed whenever the network link between the Argo CD control plane and the managed cluster may be a little flaky.

# Motivation

Improve stability on high latency networks and don't fail operations on the first hiccup. 

# Proposal

On certain, non-permanent errors when accessing remote Kubernetes API endpoints, we should retry the request when it fails.

### PR DESCRIPTION
**Pull request by aborilov**
This PR introduces a retry mechanism to the Kubernetes client used by ArgoCD. This change aims to enhance the resilience of our interactions with the Kubernetes API server.

Introduced new environment variables `ARGOCD_K8SCLIENT_RETRY_MAX` and `ARGOCD_K8SCLIENT_RETRY_BASE_BACKOFF`. The retry mechanism will be activated only when `ARGOCD_K8SCLIENT_RETRY_MAX` is specified, allowing users to opt-in to this feature.

Fixes: https://github.com/argoproj/argo-cd/issues/7692

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

### CODE REVIEW

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -2891,6 +2891,12 @@ func SetK8SConfigDefaults(config *rest.Config) error {
 	config.Timeout = K8sServerSideTimeout
 
 	config.Transport = tr
+	maxRetries := env.ParseInt64FromEnv(utilhttp.EnvRetryMax, 0, 1, math.MaxInt64)
+	if maxRetries > 0 {
+		backoffDurationMS := env.ParseInt64FromEnv(utilhttp.EnvRetryBaseBackoff, 100, 1, math.MaxInt64)
+		backoffDuration := time.Duration(backoffDurationMS) * time.Millisecond
+		config.WrapTransport = utilhttp.WithRetry(maxRetries, backoffDuration)
+	}
```
Comment (gdsoumya) [inline pkg/apis/application/v1alpha1/types.go:2899]: For better discovery, we should add these env vars to the cli [here](https://github.com/argoproj/argo-cd/blob/c47ea326cb5e731e30b2e1ec9b41c38fe53b275a/cmd/argocd-application-controller/commands/argocd_application_controller.go#L190) and pass them down. Also might be a good idea to document these in env in the docs.

**Review by gdsoumya — CHANGES_REQUESTED** [review]

File: pkg/apis/application/v1alpha1/types.go (hunk 1 above)
Comment (gdsoumya) [inline pkg/apis/application/v1alpha1/types.go:2899]: Might be a good idea to document these in env in the docs, for visibility.

**Review by gdsoumya — CHANGES_REQUESTED** [review]

File: pkg/apis/application/v1alpha1/types.go (hunk 1 above)
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go:2899]: Agree: I would add to `argocd-cmd-params-cm` configmap and add optional mounts in controller/api server deployments. Example PR https://github.com/argoproj/argo-cd/pull/15646/files

Comment (aborilov) [reply pkg/apis/application/v1alpha1/types.go:2900]: done

**Review by alexmt — APPROVED** [review]
LGTM!

**Review by gdsoumya — APPROVED** [review]
LGTM

**Pull request merged**